File: src/agents/alerter/src/farmos/commits/commit-harvest.js
'use strict';

// Phase 40 B7.5 harvest (D-03b). Option A hybrid shape (2026-05-14):
//   * Bag assets ARE fungi assets: fungi_type=<strain>, fungi_xing='fruit',
//     parent=source blocks (C4 lineage).
//   * NO harvest_batch fungi asset under the new schema -- 'batch' is not
//     a valid fungi_xing value, and the harvest log itself bundles the
//     bags + source blocks together. harvest_batch_name (if present) is
//     preserved in the log notes for human-readable lineage.
//   * Missing source block aborts BEFORE any farmOS write. Bag QR
//     collision is terminal.
//
// Strain resolution: drafts don't carry an explicit strain field;
// extract from harvest_batch_name (HBATCH-...-{STRAIN}-...) or fall back
// to draft.strain / draft.species_code. Document missing-strain as a
// terminal failure -- caller must include it.
//
// Cross-ref: .planning/notes/2026-05-14-reply-from-farmos-fungi-schema.md

const assets = require('../assets');
const logs = require('../logs');
const qrMod = require('../qr');

// HBATCH-2026-05-13-DT-001 -> 'DT'.  Matches B5 strain codes (2-4 chars).
const HBATCH_STRAIN_RE = /-([A-Z]{2,4})-[0-9]+$/;

function resolveStrain(dj) {
  if (dj.strain) return dj.strain;
  if (dj.fungi_type) return dj.fungi_type;
  if (dj.species_code) return dj.species_code;
  if (dj.species) return dj.species;
  if (dj.harvest_batch_name) {
    const m = HBATCH_STRAIN_RE.exec(dj.harvest_batch_name);
    if (m) return m[1];
  }
  return null;
}

async function commitHarvest(client, draft, ctx) {
  const dj = draft.draft_json || {};
  const draftId = draft.id;
  const timestamp = typeof dj.timestamp === 'number' ? dj.timestamp : (Date.now() / 1000);

  const sourceQrs = Array.isArray(dj.source_qr_codes) ? dj.source_qr_codes : [];
  const bags = Array.isArray(dj.bags) ? dj.bags : [];

  // 1. Resolve source blocks. Pre-check ALL before any write.
  const sourceIds = [];
  for (const qr of sourceQrs) {
    const r = await qrMod.resolveQr(client, qr);
    if (!r.found || !r.assetId) {
      return { ok: false, reason: 'missing_source_block' };
    }
    sourceIds.push(r.assetId);
  }
  if (sourceIds.length === 0) {
    return { ok: false, reason: 'missing_source_block' };
  }

  // 1b. Pre-check bag QRs are unbound (no collision).
  for (const bag of bags) {
    if (!bag || !bag.qr_code) continue;
    const r = await qrMod.resolveQr(client, bag.qr_code);
    if (r.found && r.assetId) {
      return { ok: false, reason: 'qr_already_bound_for_bag' };
    }
  }

  // 2. Strain required for bag fungi_type.
  const strain = resolveStrain(dj);
  if (!strain) return { ok: false, reason: 'missing_strain' };

  // 3. Create bag assets (parents = source blocks).
  const batchName = dj.harvest_batch_name; // labelling only; not an asset.
  const bagIds = [];
  for (const bag of bags) {
    const bagName = bag.name || `${batchName || 'harvest'}-bag-${bagIds.length + 1}`;
    // Phase 51 UPSERT-01: bag asset goes through upsertFungiAsset so re-runs
    // converge instead of duplicating bags. Bag QR pre-check above already
    // ensures the QR slot is free; upsert semantics make name-based idempotency
    // safe for accidental replay.
    const bagRes = await assets.upsertFungiAsset(client, {
      name: bagName,
      parentIds: sourceIds,
      fungiTypeName: strain,
      fungiXingName: 'fruit',
      qrCodes: bag.qr_code ? [bag.qr_code] : [],
      draftId,
    });
    if (!bagRes.ok) return { ok: false, reason: bagRes.reason || 'bag_upsert_failed', http_status: bagRes.http_status, asset_ids: [...bagIds] };
    bagIds.push(bagRes.assetId);
  }

  // 4. Harvest log: order = source blocks, bags.
  const assetIds = [...sourceIds, ...bagIds];
  const weightLines = bags
    .map((b, i) => `bag${i + 1}: ${b.weight_grams != null ? b.weight_grams + 'g' : 'n/a'}`)
    .join('\n');
  const noteParts = [];
  if (dj.notes) noteParts.push(dj.notes);
  if (batchName) noteParts.push('harvest_batch: ' + batchName);
  if (weightLines) noteParts.push('Weights:\n' + weightLines);
  const notes = noteParts.join('\n');
  const name = `harvest ${new Date(timestamp * 1000).toISOString().slice(0, 10)}`;
  const logRes = await logs.createLog(client, 'harvest', {
    name, timestamp, assetIds, notes, draftId,
  });
  if (!logRes.ok) return { ok: false, reason: logRes.reason, http_status: logRes.http_status, asset_ids: [...bagIds] };

  return {
    ok: true,
    asset_ids: [...bagIds],
    log_ids: [logRes.logId],
    file_ids: [],
    http_status: logRes.http_status,
  };
}

module.exports = commitHarvest;

File: src/farm-agent/farm_agent/farmos/commits/commit_harvest.py
"""farm_agent/farmos/commits/commit_harvest.py -- Harvest commit handler.

Faithful Python port of src/agents/alerter/src/farmos/commits/commit-harvest.js
(Phase 40 B7.5 D-03b / Phase 62-08).

Option A hybrid shape (2026-05-14):
  * Bag assets ARE fungi assets: fungi_type=<strain>, fungi_xing='fruit',
    parent=source blocks (C4 lineage).
  * NO harvest_batch fungi asset under the new schema -- 'batch' is not
    a valid fungi_xing value, and the harvest log itself bundles the
    bags + source blocks together. harvest_batch_name (if present) is
    preserved in the log notes for human-readable lineage.
  * Missing source block aborts BEFORE any farmOS write. Bag QR
    collision is terminal.

Strain resolution: extract from harvest_batch_name (HBATCH-...-{STRAIN}-...)
or fall back to draft.strain / draft.species_code. Missing strain = terminal.

Returns the uniform commit envelope:
  {"ok": bool, "asset_ids": list, "log_ids": list, "file_ids": list,
   "http_status": int|None, "reason": str|None}

ASCII-only. No em-dashes. Never-throws at the handler level.
"""
from __future__ import annotations

import re
import time
from datetime import datetime, timezone

from farm_agent.farmos.assets import upsert_fungi_asset
from farm_agent.farmos.logs import create_log
from farm_agent.farmos.qr import resolve_qr

# HBATCH-2026-05-13-DT-001 -> 'DT'. Matches B5 strain codes (2-4 uppercase chars).
_HBATCH_STRAIN_RE = re.compile(r"-([A-Z]{2,4})-[0-9]+$")


def _resolve_strain(dj: dict) -> str | None:
    """Extract strain code from draft_json fields. Mirrors resolveStrain() in JS."""
    if dj.get("strain"):
        return dj["strain"]
    if dj.get("fungi_type"):
        return dj["fungi_type"]
    if dj.get("species_code"):
        return dj["species_code"]
    if dj.get("species"):
        return dj["species"]
    if dj.get("harvest_batch_name"):
        m = _HBATCH_STRAIN_RE.search(dj["harvest_batch_name"])
        if m:
            return m.group(1)
    return None


async def commit_harvest(client: dict, draft: dict, ctx: dict | None = None) -> dict:
    """Create bag assets and a harvest log for the given draft.

    Signature: async commit_harvest(client, draft, ctx=None) -> envelope dict.

    Port of commitHarvest() from commit-harvest.js lines 39-117.
    """
    dj = draft.get("draft_json") or {}
    draft_id = draft.get("id", "")
    timestamp = dj["timestamp"] if isinstance(dj.get("timestamp"), (int, float)) else time.time()

    source_qrs = dj.get("source_qr_codes") if isinstance(dj.get("source_qr_codes"), list) else []
    bags = dj.get("bags") if isinstance(dj.get("bags"), list) else []

    # 1. Resolve source blocks. Pre-check ALL before any write.
    source_ids = []
    for qr in source_qrs:
        r = await resolve_qr(client, qr)
        if not r.get("found") or not r.get("asset_id"):
            return {"ok": False, "reason": "missing_source_block"}
        source_ids.append(r["asset_id"])

    if not source_ids:
        return {"ok": False, "reason": "missing_source_block"}

    # 1b. Pre-check bag QRs are unbound (no collision).
    for bag in bags:
        if not bag or not bag.get("qr_code"):
            continue
        r = await resolve_qr(client, bag["qr_code"])
        if r.get("found") and r.get("asset_id"):
            return {"ok": False, "reason": "qr_already_bound_for_bag"}

    # 2. Strain required for bag fungi_type.
    strain = _resolve_strain(dj)
    if not strain:
        return {"ok": False, "reason": "missing_strain"}

    # 3. Create bag assets (parents = source blocks).
    batch_name = dj.get("harvest_batch_name")  # labelling only; not an asset.
    bag_ids = []
    for bag in bags:
        bag_name = bag.get("name") or (
            (batch_name or "harvest") + "-bag-" + str(len(bag_ids) + 1)
        )
        bag_res = await upsert_fungi_asset(client, {
            "name": bag_name,
            "parent_ids": source_ids,
            "fungi_type_name": strain,
            "fungi_xing_name": "fruit",
            "qr_codes": [bag["qr_code"]] if bag.get("qr_code") else [],
            "draft_id": draft_id,
        })
        if not bag_res.get("ok"):
            return {
                "ok": False,
                "reason": bag_res.get("reason") or "bag_upsert_failed",
                "http_status": bag_res.get("http_status"),
                "asset_ids": list(bag_ids),
            }
        bag_ids.append(bag_res["asset_id"])

    # 4. Harvest log: order = source blocks, bags.
    all_asset_ids = [*source_ids, *bag_ids]
    weight_lines = "\n".join(
        "bag" + str(i + 1) + ": " + (str(b.get("weight_grams")) + "g" if b.get("weight_grams") is not None else "n/a")
        for i, b in enumerate(bags)
    )
    note_parts = []
    if dj.get("notes"):
        note_parts.append(dj["notes"])
    if batch_name:
        note_parts.append("harvest_batch: " + batch_name)
    if weight_lines:
        note_parts.append("Weights:\n" + weight_lines)
    notes = "\n".join(note_parts)

    name = "harvest " + datetime.fromtimestamp(timestamp, tz=timezone.utc).strftime("%Y-%m-%d")
    log_res = await create_log(client, "harvest", {
        "name": name,
        "timestamp": timestamp,
        "asset_ids": all_asset_ids,
        "notes": notes,
        "draft_id": draft_id,
    })
    if not log_res.get("ok"):
        return {
            "ok": False,
            "reason": log_res.get("reason"),
            "http_status": log_res.get("http_status"),
            "asset_ids": list(bag_ids),
        }
    return {
        "ok": True,
        "asset_ids": list(bag_ids),
        "log_ids": [log_res["log_id"]],
        "file_ids": [],
        "http_status": log_res.get("http_status"),
    }
